Feature/iss1 patchwork integration

dchan opened the pull request:
This implements the basic Patchwork API [1] as a mcv module. Tested on python 2.7.6 with supplied requirements.txt

There is no support for the GET functionalities in this PR.

[1] - https://patchworksecurity.com/docs/

elliot42 commented:
Not compliant with pep8, please run `pep8` against the file: https://pypi.python.org/pypi/pep8

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,150 @@
+"""Patchwork"""
```
Description + website link would be useful here

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,150 @@
+PATCHWORK_API_ENDPOINT = os.getenv('PATCHWORK_API_ENDPOINT',
+                                   "https://api.patchworksecurity.com/api/v1")
+DEFAULT_UUID_PATH = '/etc/patchwork/uuid'
+DPKG_QUERY_CMD = ["/usr/bin/dpkg-query", "-W",
```
This could equally go in the `mcv.apt` module as a Python function, and then be called from this module.  Not a blocker.

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,150 @@
+        with open(uuid_path, 'r') as f:
+            machine_data = json.load(f)
+
+        return machine_data
```
may as well just `return json.load(f)`?

dchan replied:
yea, that should just be a return

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,150 @@
+
+        return machine_data
+
+    _mkdirs(uuid_path)
```
Would expect `_mkdirs` to be defined above this function if this function depends upon it

elliot42 replied:
Or, it's only called from one place so may as well be inline?

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,150 @@
+    if not os.path.exists(dirs):
+        os.makedirs(dirs)
+
+def _get_os_info(path='/etc/lsb-release'):
```
Ideally this would probably go in like `mcv.distro` or something, I think, so we'd have a general purpose distro identifier function.

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,150 @@
+def _get_os_info(path='/etc/lsb-release'):
+    """Parse machine release information
+
+    This doesn't work for all Linux distributions
```
Probably worth mentioning what it *does* work for currently

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,150 @@
+        PATCHWORK_API_ENDPOINT,
+        "machine"])
+
+    machine_data = _post(api_key=api_key, url=url, data=os_data)
```
Would expect `_post`, `_make_request` etc to be defined above this function if this function depends upon them

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,150 @@
+def _save_uuid(uuid_path, data):
+    """Save machine UUID metadata to file"""
+    if 'uuid' not in data:
+        return
```
This seems more like an assert/throw situation more than a silently do nothing?  Correct me if I'm off here.

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,150 @@
+
+    return data['uuid']
+
+def _save_uuid(uuid_path, data):
```
Feel like this may as well be inline since it's only used in one place

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,150 @@
+
+    return os_data
+
+def _get_uuid(uuid_path):
```
This could probably be inline if it's only used in one place

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,150 @@
+
+    return machine_data
+
+def update(api_key, packages, uuid=None, uuid_path=DEFAULT_UUID_PATH):
```
If `packages` needs a specific structure (it looks like it's list of 3-tuples? `[[status, package, version], ...]`) probably worth mentioning in docstring

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,157 @@
+
+    data = []
+
+    for status, package, version in packages:
```
This seems like it can be accomplished more idiomatically with a list comprehension

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,157 @@
+
+
+def gather_packages():
+    """Get the list of packages installed on this machine"""
```
Probably useful to document the return value structure, since the output of this is used as the input of `update`

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,157 @@
+    return _post(api_key=api_key, url=url, data=data)
+
+
+def gather_packages():
```
Ideally I think this would go in `mcv.apt` (and then one day you'd need to figure out how to generalize it across apt, yum, etc.)

elliot42 commented on mcv/patchwork.py:
```diff
@@ -0,0 +1,157 @@
+    return [line.split('\t') for line in output.splitlines() if line]
+
+
+def _make_request(api_key, url, method='get', params=None, data=None):
```
As mentioned above would expect this to be declared before all the stuff that uses it

elliot42 commented:
Some organizational/cleanup feedback, otherwise LGTM.  Will wait for a comment that the next rev is ready.

dchan commented:
I reorganized the file and removed unnecessary functions. I added mcv.os for os info gathering, though it only works on for Ubuntu right now. The code should be simpler than the previous version.

elliot42 commented:
Looks good thank you!!
